changefeedccl: protect system tables with their own pts record

aerfrei opened the pull request:
Previously, we were relying on each pts record that protects
a user table to also protect system tables. This would mean
that we'd protect the system tables once for each table.

Now, we have a dedicated pts record for system tables that
protects back to the highwater and do not include the system
table protections in the table specific records. We only
protect the system tables once.

Fixes: #152446
Epic: CRDB-1421

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152448)
<!-- Reviewable:end -->

KeithCh commented on pkg/ccl/changefeedccl/changefeed_test.go:
Flagging this

KeithCh commented on pkg/ccl/changefeedccl/protected_timestamps_test.go:
```diff
@@ -1054,7 +1074,7 @@ func TestChangefeedProtectedTimestampUpdateForMultipleTables(t *testing.T) {
 	})
 
-	cdcTest(t, testFn, feedTestForceSink("kafka"), withTxnRetries)
+	cdcTest(t, testFn, feedTestOmitSinks("sinkless"), withTxnRetries)
```
Nit: `feedTestEnterpriseSinks`

aerfrei commented on pkg/ccl/changefeedccl/cdcprogresspb/progress.proto:
```diff
@@ -22,9 +22,24 @@ message ResolvedTables {
 message ProtectedTimestampRecords {
-  map<uint32, bytes> protected_timestamp_records = 1 [
+  // PrimaryRecord protects tables that are not lagging.
+  bytes primary_record = 1 [
```
Before, we were storing this record in the changefeed progress just like the single-record (non-per table) feeds. Especially now that some of those pts record targets are made with system tables and some are not, I think it helps to store them in different places. It probably could be separated from this PR, but I don't think it would have a large impact on the diff size.

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -1832,12 +1832,29 @@ func (b *changefeedResumer) OnFailOrCancel(
+		for _, perTableRecord := range ptsEntries.PerTableRecords {
+			maybeCleanUpProtectedTimestamp(perTableRecord)
+		}
+		return nil
```
Should we also delete the file after cleaning up the records?

andyyang890 commented on pkg/ccl/changefeedccl/protected_timestamps_test.go:
```diff
@@ -1150,12 +1173,13 @@ func TestChangefeedPerTableProtectedTimestampProgression(t *testing.T) {
 							return errors.Newf("expected PTS record for table %d", tableID)
```
Could we update this test to assert that the records all have different IDs as well?

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -12073,8 +12075,25 @@ func TestChangefeedProtectedTimestampUpdate(t *testing.T) {
+				return err
+			}
+
+			// If the file exists, the records should be nil
```
Why would the file exist?

aerfrei committed:
changefeedccl: protect system tables with their own pts record

Previously, we were relying on each pts record that protects
a user table to also protect system tables. This would mean
that we'd protect the system tables once for each table.

Now, we have a dedicated pts record for system tables that
protects back to the highwater and do not include the system
table protections in the table specific records. We only
protect the system tables once.

Fixes: #152446
Epic: CRDB-1421

Release note: None

aerfrei commented on pkg/ccl/changefeedccl/protected_timestamps_test.go:
```diff
@@ -799,25 +809,61 @@ func TestChangefeedMigratesProtectedTimestampTargets(t *testing.T) {
 			})
 		}
 
-		// Wipe out the targets from the changefeed PTS record, simulating an old-style PTS record.
```
I changed this comment because it's somewhat inaccurate/unclear. I think the "old-style" pts record actually refers to pts records that didn't have targets at all, but this is really testing the scenario where the targets are incomplete and are missing specific system tables.

aerfrei commented on pkg/ccl/changefeedccl/protected_timestamps.go:
```diff
@@ -31,14 +34,52 @@ func createProtectedTimestampRecord(
+	ctx context.Context, codec keys.SQLCodec, jobID jobspb.JobID, resolved hlc.Timestamp,
+) *ptpb.Record {
+	ptsID := uuid.MakeV4()
+	deprecatedSpansToProtect := makeSpansToProtect(codec, changefeedbase.Targets{})
```
This is deprecated and should be removed. We should not create new functions that create deprecated spans to protect if they are deprecated. I think I can likely pass an empty slice here, but was trying to maintain consistency until we can actually fully remove.

Comment says
```
// TODO(adityamaru): In 22.2 stop passing `deprecatedSpans` since PTS records
// will stop protecting key spans.
```

log-head commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -1973,13 +1989,13 @@ func (b *changefeedResumer) OnFailOrCancel(
 		}
-
-		if len(ptsEntries.TableRecords) == 0 {
+		if len(ptsEntries.TableRecords) == 0 && ptsEntries.SystemTablesRecord == uuid.Nil {
```
Could you rename ptsEntries here to something like perTablePtsEntries? To make it more clear that non-per-table protected changefeeds will return at this point? And everything after is for per-table protections

aerfrei replied:
I left a comment to clarify this.

log-head commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -2124,14 +2135,63 @@ func (cf *changeFrontier) createPerTablePTSTargets(
+		return false, err
+	}
+
+	if !makeSystemTablesTargetToProtect().Equal(rec.Target) {
```
Does this only happen on a version upgrade, that the set of protected system tables changes?

aerfrei replied:
Yes, this should happen if we change which system tables are protected for changefeeds, which we may do in a minor version upgrade. Otherwise, it would be a bug like this one: https://github.com/cockroachdb/cockroach/issues/151690. That is only known to happen for the old combined style pts on changefeeds that watch multiple tables.

log-head reviewed (approved):
lgtm!

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -2124,14 +2135,63 @@ func (cf *changeFrontier) createPerTablePTSTargets(
+	}
+
+	if !makeSystemTablesTargetToProtect().Equal(rec.Target) {
+		if preservePTSTargets := cf.knobs.PreservePTSTargets != nil && cf.knobs.PreservePTSTargets(); preservePTSTargets {
```
nit: unnecessary bool var

aerfrei commented:
> ## Potential Bug(s) Detected
These are saying I haven't dealt with the backwards compatibility of the protobuf change. It doesn't seem to know that none of these protobufs exist to be backwards compatible with.

andyyang890 commented on pkg/ccl/changefeedccl/protected_timestamps.go:
```diff
@@ -31,14 +34,50 @@ func createProtectedTimestampRecord(
+
+	log.VEventf(ctx, 2, "creating protected timestamp for system tables %v at %v", ptsID, resolved)
+	return jobsprotectedts.MakeRecord(
+		ptsID, int64(jobID), resolved, nil, jobsprotectedts.Jobs, makeSystemTablesTargetToProtect(),
```
nit: add comment for the `nil` param name https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+Golang+coding+guidelines#Avoid-Naked-Parameters

ditto for `createUserTablesProtectedTimestampRecord` below

andyyang890 reviewed (approved):
Nice, thanks for pushing this over the finish line
